filters/imagefilter.go

```diff
@@ -43,65 +41,74 @@ func init() {
 
 type ImageFilter interface {
 	CreateOptions(image *bimg.Image) (*bimg.Options, error)
+	CanBeMerged(other *bimg.Options, self *bimg.Options) bool
+	Merge(other *bimg.Options, self *bimg.Options) *bimg.Options
 }
 
 func HandleImageResponse(ctx filters.FilterContext, f ImageFilter) {
-	rsp := ctx.Response()
-
-	defer rsp.Body.Close()
-
-	rsp.Header.Del("Content-Length")
-
-	r, err := handleImageTransform(rsp.Body, f)
+	image := ctx.StateBag()[SkropImage].(*bimg.Image)
 
+	opt, err := f.CreateOptions(image)
 	if err != nil {
-		log.Error("failed to process image ", err.Error())
+		log.Error("Failed to create options ", err.Error())
 		ctx.Serve(&http.Response{
-			StatusCode:http.StatusInternalServerError,
-			Body: ioutil.NopCloser(bytes.NewBufferString(err.Error())),
+			StatusCode: http.StatusInternalServerError,
+			Body:       ioutil.NopCloser(bytes.NewBufferString(err.Error())),
 		})
 		return
 	}
 
-	rsp.Body = ioutil.NopCloser(r)
-}
+	opts := ctx.StateBag()[SkropOptions].(*bimg.Options)
 
-func handleImageTransform(r io.Reader, f ImageFilter) (io.Reader, error) {
-	var err error
-	imageBytes, err := ioutil.ReadAll(r)
+	if f.CanBeMerged(opts, opt) {
+		ctx.StateBag()[SkropOptions] = f.Merge(opts, opt)
+		log.Debug("Filter ", f, " merged in ", ctx.StateBag()[SkropOptions])
+		return
+	}
 
+	//transform the image
+	buf, err := transformImage(image, opts)
```

Comment:
Yes, what I was trying to explain is that, the CreateOption function takes an image as input, and returns a bimg.Option on how to change that image. and this is executed in a chain.
For example the filter resizeLongerEdge takes an image, check if the image is landscape or portrait and if it is landscape set the width to a value, if it is a portrait it sets the height.
Consider the following eskip:
```
small: Path("/images/S/:image")
  -> longerEdgeResize(800)
  -> crop(200, 500)
  -> "http://localhost:9090";
```
The backend returns a landscape image.
The first filter returns a bimg.Option with width=200 and height=500 and the option is saved in the statebag
The second returns CanBeMerged=false
if I do not process the image here, I call longerEdgeResize and as input I give a landscape image. But this is wrong, because the previous filter crop it to a portrait image.

That's why, in case two options cannot be merged, I have to first process the image and then to call the CreateOption of the filter.